base/commandLine.cpp

```diff
@@ -1631,6 +1664,13 @@ bool processSettings(Common::String &command, Common::StringMap &settings, Commo
 	} else if (command == "add") {
 		addGames(settings["path"], gameOption.engineId, gameOption.gameId, settings["recursive"] == "true");
 		return true;
+	} else if (command == "md5") {
+		char* end;
+		Common::String filename = settings["md5-path"];
+		Common::Path Filename(filename, '/');
+		Common::FSNode path(Filename);
+		calcMD5(path, strtol(settings["md5-length"].c_str(), &end, 10));
```

Comment:
Does that work when no length is specified on the command line?

Also I am wondering what would be the best default. Should it be 0 (use the whole file), or 5000 (what most engines are using)?

Reply:
If no length is passed on the command line, the MD5 hash of the whole file is calculated. That also happens when ```--md5-length=0``` is passed.

And I have some doubts about the ```--md5-length``` option. Should we keep it so that user can specify number of bytes or shall we ask them the engine of the game they want the MD5 hash of? We can ask them to enter the engine name as given by the ```--list-engines``` command and then we can use the data [here](https://wiki.scummvm.org/index.php/Reporting_unknown_MD5_checksums) to find the MD5 hash of a game for a given engine.

We can also give both of the options, with engine name overriding ```--md5-length``` if both options are passed by the user. Please tell if this should be done or not.

Reply:
I was expecting an error as the `HashMap::operator[] const` results in an error (in debug build) when the key is not present in the map. But here we have a non-const reference to a HashMap, so indeed instead the operator will insert an empty string to the map.

That is still not working as expected though since:
 - If `--md5-path` was not used we end up with an empty  path, which for me results in trying to apply the command on the current directory (and in the following error: `FSNode::createReadStream: 'scummvm' is a directory!`).
 - If `--md5-length` was not used, we end up calling `strtol()` on an empty string. This work, but may not be portable (the man page for `strtol()` tells me `If no conversion could be performed, 0 is returned and the global variable errno is set to EINVAL (the feature is not portable across all platforms).`).

So I think it would be clearer (and maybe necessary) to check if the value is present and use an explicit default.
For the path you seem to have documented `./scummvm` as a default, so it would look like this:
```c++
Common::String filename = settings.getValOrDefault("md5-path", "scummvm");
```
For the length you could also use `String:: asUint64()` instead of `strtol()`:
````c++
long int md5Length = 0;
if (settings.contains("md5-length"))
	md5Length = settings["md5-length"].asUint64();
````

> And I have some doubts about the --md5-length option. Should we keep it so that user can specify number of bytes or shall we ask them the engine of the game they want the MD5 hash of? We can ask them to enter the engine name as given by the --list-engines command and then we can use the data [here](https://wiki.scummvm.org/index.php/Reporting_unknown_MD5_checksums) to find the MD5 hash of a game for a given engine.

The `AdvancedMetaEngineDetection`, a specialisation of `MetaEngineDetection` that uses md5 tables for detection, has the information of how many bytes to use. Currently this is not exposed publicly, but we could easily add a public getting function to the class to get the info. So if you want to add a `--md5-engine=<engine>` option, it should be querying that instead of hardcoding values based on what is documented in the wiki.

And I don't think such an option should replace `--md5-length`, but it could indeed be added as an alternative. After all we may want to compute the md5 for reasons other than detection (e.g. to check if a file is corrupted).